l10n_it_fatturapa_out_sp/tests/fatturapa_common.py

```diff
@@ -0,0 +1,249 @@
+import base64
+import os
+import shutil
+import tempfile
+
+from lxml import etree
+
+from odoo.modules.module import get_module_resource
+
+from odoo.addons.account.tests.common import AccountTestInvoicingCommon
+from odoo.addons.mail.tests.common import mail_new_test_user
+
+
+class FatturaPACommon(AccountTestInvoicingCommon):
+    def setUp(self):
+        super(FatturaPACommon, self).setUp()
+        # used to be in AccountTestUsers
```

Comment:
Eh, questo fa parte del famoso refactoring dei test. Il problema è che i vari l10n_fatturapa_*/tests/fatturapa_common.py non sono fatti per essere condivisi. As es.:

```diff -u l10n_it_fatturapa_out/tests/fatturapa_common.py l10n_it_fatturapa_out_sp/tests/fatturapa_common.py
--- l10n_it_fatturapa_out/tests/fatturapa_common.py	2021-06-09 11:17:07.192960103 +0200
+++ l10n_it_fatturapa_out_sp/tests/fatturapa_common.py	2021-07-30 12:12:04.230385351 +0200
@@ -116,6 +116,7 @@
                 "description": tax.description,
                 "amount": tax.amount,
                 "type_tax_use": tax.type_tax_use,
+                "payability": tax.payability,
                 "invoice_repartition_line_ids": invoice_rpls,
                 "refund_repartition_line_ids": refund_rpls,
             }
@@ -139,6 +140,32 @@
         self.res_partner_fatturapa_4 = self.env.ref(
             "l10n_it_fatturapa.res_partner_fatturapa_4"
         )
+        self.fiscal_position_sp = self.env["account.fiscal.position"].create(
+            {
+                "name": "Split payment",
+                "split_payment": True,
+                "tax_ids": [
+                    (
+                        0,
+                        0,
+                        {
+                            "tax_src_id": self.tax_22.id,
+                            "tax_dest_id": self.tax_22_SP.id,
+                        },
+                    )
+                ],
+            }
+        )
+        self.company.sp_account_id = self.env["account.account"].search(
+            [
+                (
+                    "user_type_id",
+                    "=",
+                    self.env.ref("account.data_account_type_current_assets").id,
+                )
+            ],
+            limit=1,
+        )
         self.EUR = self.env.ref("base.EUR")
         # United Arab Emirates currency
         self.AED = self.env.ref("base.AED")
@@ -146,8 +173,6 @@
             "UPDATE res_company SET currency_id = %s WHERE id = %s",
             [self.EUR.id, self.company.id],
         )
-        # Otherwise self.company in cache could keep the old wrong value USD
-        self.company.refresh()
 
     def AttachFileToInvoice(self, InvoiceId, filename):
         self.fatturapa_attach.create(
@@ -197,6 +222,8 @@
         test_fatt_content = base64.decodebytes(test_fatt_data)
         test_fatt = etree.fromstring(test_fatt_content, parser)
         xml = etree.fromstring(xml_content, parser)
+        with open("/tmp/my.xml", "wb") as o:
+            o.write(etree.tostring(xml))
         self.assertEqual(etree.tostring(test_fatt), etree.tostring(xml))
 
     def getFilePath(self, filepath):
@@ -208,7 +235,7 @@
 
     def getAttachment(self, name, module_name=None):
         if module_name is None:
-            module_name = "l10n_it_fatturapa_out"
+            module_name = "l10n_it_fatturapa_out_sp"
         path = get_module_resource(module_name, "tests", "data", "attah_base.pdf")
         currDir = os.path.dirname(path)
         new_file = "{}/{}".format(currDir, name)
@@ -217,36 +244,6 @@
 
     def getFile(self, filename, module_name=None):
         if module_name is None:
-            module_name = "l10n_it_fatturapa_out"
+            module_name = "l10n_it_fatturapa_out_sp"
         path = get_module_resource(module_name, "tests", "data", filename)
         return self.getFilePath(path)
-
-    def _create_invoice(self):
-        return self.invoice_model.create(
-            {
-                "name": "Test Invoice",
-                "journal_id": self.sales_journal.id,
-                "partner_id": self.res_partner_fatturapa_0.id,
-                "move_type": "out_invoice",
-                "invoice_line_ids": [
-                    (
-                        0,
-                        0,
-                        {
-                            "account_id": self.a_sale.id,
-                            "product_id": self.product_product_10.id,
-                            "name": self.product_product_10.name,
-                            "quantity": 1,
-                            "price_unit": 1,
-                            "tax_ids": [(6, 0, {self.tax_22.id})],
-                        },
-                    ),
-                ],
-            }
-        )
-
-    def _create_e_invoice(self):
-        invoice = self._create_invoice()
-        invoice._post()
-        action = self.run_wizard(invoice.id)
-        return self.env[action["res_model"]].browse(action["res_id"])
-
```

sono chiaramente parenti, prob. alcune cose sono state aggiunte in fatturapa_out in seguito che andrebbero riportate di qua. Ma il problema è che il nome del modulo è cablato. Andrebbe riscritto e unificato in modo che tutti i moduli possano importare la classe e customizzarla.  E dovrebbe partire da fatturapa -> fatturapa_out -> fatturapa_out_sp customizzando solo il necessario.

Comunque vedo se riesco a razionalizzarlo un po' per questo modulo.

Reply:
Ok, fatto, meno peggio di quanto pensassi. Ci sono un paio di cose che non servono per i test di l10n_it_fatturapa_out ma servono per questo modulo, tuttavia per pulizia estetica andrebbero prob. spostati in l10n_it_fatturapa_out.